code/parse/sexp.cpp

```diff
@@ -11737,69 +11737,106 @@ void sexp_set_subsystem_strength(int n)
 }
 
 // destroys a subsystem without explosions
-void sexp_destroy_subsys_instantly (int n)
+void sexp_destroy_subsys_instantly(int n)
 {
 	char *subsystem;
-	int	shipnum, index, generic_type;
+	int	shipnum, subsys_index, generic_type;
 	ship *shipp;
-	ship_subsys *ss = NULL;
-	
-	// if MULTIPLAYER bail
-	if (Game_mode & GM_MULTIPLAYER) {
-		return;
-	}
+	ship_subsys *ss;
 
 	shipnum = ship_name_lookup(CTEXT(n));
 	// if no ship, then return immediately.
-	if ( shipnum == -1 ){
+	if (shipnum < 0) {
 		return;
 	}
 
 	shipp = &Ships[shipnum];
 	n = CDR(n);
-	
-	//Process subsystems
-	while(n != -1)
+
+	if (MULTIPLAYER_MASTER)
+	{
+		Current_sexp_network_packet.start_callback();
+		Current_sexp_network_packet.send_ship(shipp);
+	}
+
+	for ( ; n >= 0; n = CDR(n))
 	{
 		subsystem = CTEXT(n);
-		if ( !stricmp( subsystem, SEXP_HULL_STRING) || !stricmp( subsystem, SEXP_SIM_HULL_STRING)){
-			n = CDR(n);
+
+		// use destroy-instantly if we want to do this
+		if (!stricmp(subsystem, SEXP_HULL_STRING) || !stricmp(subsystem, SEXP_SIM_HULL_STRING))
 			continue;
-		}
+
 		// deal with generic subsystems
 		generic_type = get_generic_subsys(subsystem);
-		if (generic_type) {
-			for (ss = GET_FIRST(&shipp->subsys_list); ss != END_OF_LIST(&shipp->subsys_list); ss = GET_NEXT(ss)) {
-				if (generic_type == ss->system_info->type) {
+		if (generic_type != SUBSYSTEM_NONE)
+		{
+			for (ss = GET_FIRST(&shipp->subsys_list); ss != END_OF_LIST(&shipp->subsys_list); ss = GET_NEXT(ss))
+			{
+				if (generic_type == ss->system_info->type)
+				{
 					// do destruction stuff
 					ss->current_hits = 0;
 					ship_recalc_subsys_strength(shipp);
-					do_subobj_destroyed_stuff(shipp, ss, NULL, true);
+					do_subobj_destroyed_stuff(shipp, ss, nullptr, true);
+
+					if (MULTIPLAYER_MASTER)
+					{
+						subsys_index = ship_get_subsys_index(shipp, ss);
+						Assert(subsys_index >= 0);
+						Current_sexp_network_packet.send_int(subsys_index);
+					}
 				}
 			}
 		}
+		// normal subsystems
 		else
 		{
-			// normal subsystems
 			ss = ship_get_subsys(shipp, subsystem);
-			index = ship_get_subsys_index(shipp, subsystem, 1);
-			if ( index == -1 ) {
-				nprintf(("Warning", "Couldn't find subsystem %s on ship %s for destroy-subsys-instantly\n", subsystem, shipp->ship_name));
-				n = CDR(n);
-				continue;
-			}
-			if(ss == NULL) {
-				nprintf(("Warning", "Nonexistent subsystem for index %d on ship %s for destroy-subsys-instantly\n", index, shipp->ship_name));
+			if (ss == nullptr)
+			{
+				nprintf(("Warning", "Nonexistent subsystem '%s' on ship %s for destroy-subsys-instantly\n", subsystem, shipp->ship_name));
 				n = CDR(n);
```

Comment:
Is this still correct since `n = CDR(n)` is already done by the `for` loop?

Reply:
Good call.  The original patch was a while loop, and I changed it, but I missed this.

Reply:
Should be fixed in the new commits.